arbnode/execution/executionengine.go

```diff
@@ -0,0 +1,565 @@
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/offchainlabs/nitro/arbos"
+	"github.com/offchainlabs/nitro/arbos/arbosState"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbos/l1pricing"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/staker"
+	"github.com/offchainlabs/nitro/util/sharedmetrics"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	"github.com/pkg/errors"
+)
+
+type TransactionStreamerInterface interface {
+	WriteMessageFromSequencer(pos arbutil.MessageIndex, msgWithMeta arbostypes.MessageWithMetadata) error
+	ExpectChosenSequencer() error
+	FetchBatch(batchNum uint64) ([]byte, error)
+}
+
+type ExecutionEngine struct {
+	stopwaiter.StopWaiter
+
+	bc        *core.BlockChain
+	validator *staker.BlockValidator
+	streamer  TransactionStreamerInterface
+
+	resequenceChan    chan []*arbostypes.MessageWithMetadata
+	createBlocksMutex sync.Mutex
+
+	newBlockNotifier chan struct{}
+	latestBlockMutex sync.Mutex
+	latestBlock      *types.Block
+
+	nextScheduledVersionCheck time.Time // protected by the createBlocksMutex
+
+	reorgSequencing bool
+}
+
+func NewExecutionEngine(bc *core.BlockChain) (*ExecutionEngine, error) {
+	return &ExecutionEngine{
+		bc:               bc,
+		resequenceChan:   make(chan []*arbostypes.MessageWithMetadata),
+		newBlockNotifier: make(chan struct{}, 1),
+	}, nil
+}
+
+func (s *ExecutionEngine) SetBlockValidator(validator *staker.BlockValidator) {
+	if s.Started() {
+		panic("trying to set block validator after start")
+	}
+	if s.validator != nil {
+		panic("trying to set block validator when already set")
+	}
+	s.validator = validator
+}
+
+func (s *ExecutionEngine) EnableReorgSequencing() {
+	if s.Started() {
+		panic("trying to enable reorg sequencing after start")
+	}
+	if s.reorgSequencing {
+		panic("trying to enable reorg sequencing when already set")
+	}
+	s.reorgSequencing = true
+}
+
+func (s *ExecutionEngine) SetTransactionStreamer(streamer TransactionStreamerInterface) {
+	if s.Started() {
+		panic("trying to set reorg sequencing policy after start")
+	}
+	if s.streamer != nil {
+		panic("trying to set reorg sequencing policy when already set")
+	}
+	s.streamer = streamer
+}
+
+func (s *ExecutionEngine) Reorg(count arbutil.MessageIndex, newMessages []arbostypes.MessageWithMetadata, oldMessages []*arbostypes.MessageWithMetadata) error {
+	s.createBlocksMutex.Lock()
+	successful := false
+	defer func() {
+		if !successful {
+			s.createBlocksMutex.Unlock()
+		}
+	}()
+	blockNum, err := s.MessageCountToBlockNumber(count)
+	if err != nil {
+		return err
+	}
+	// We can safely cast blockNum to a uint64 as we checked count == 0 above
+	targetBlock := s.bc.GetBlockByNumber(uint64(blockNum))
```

Comment:
MessageCountToBlockNumber could return -1 if count is 0 and this is a chain without classic blocks. However, that would be casted to a block number which wouldn't be found, and it would then log the error "reorg target block not found" with block being -1, which is a sensible error.